### PR TITLE
Pallas gpu attention- remove a potential downcast

### PR DESCRIPTION
**Pull request by luyug**
Line 397 in pallas gpu attention (backward kernel), `qk = qk.astype(q_ref.dtype)` seems to be redundant and can potentially lead to wrong results. 

Context:
```
qk = pl.dot(q, k.T)
qk = qk.astype(q_ref.dtype)
qk = qk.astype(jnp.float32)
```

Since `pl.dot` will accumulates in `float32`, the first cast is either a no-op, or a down-cast (followed by a upcast.)

In some of my experiments, this gives wrong gradients if inputs are in `(b)float16` types and the gradient norm is large.

### CODE REVIEW

**Comment by sharadmv** [comment]
Were you able to verify all the tests pass?

**Comment by luyug** [comment]
I tried the pallas test and it seems to be passing (on A100.)

```
[ RUN      ] FusedAttentionTest.test_fused_attention_bwd_batch_size=1_seq_len=384_num_heads=1_head_dim=32_causal=False_use_segment_ids=False
I0111 21:48:53.965465 140611409929792 xla_bridge.py:670] Unable to initialize backend 'rocm': NOT_FOUND: Could not find registered platform with name: "rocm". Available platform names are: CUDA
I0111 21:48:53.967653 140611409929792 xla_bridge.py:670] Unable to initialize backend 'tpu': INTERNAL: Failed to open libtpu.so: libtpu.so: cannot open shared object file: No such file or directory
[       OK ] FusedAttentionTest.test_fused_attention_bwd_batch_size=1_seq_len=384_num_heads=1_head_dim=32_causal=False_use_segment_ids=False
[ RUN      ] FusedAttentionTest.test_fused_attention_bwd_batch_size=1_seq_len=384_num_heads=1_head_dim=32_causal=False_use_segment_ids=True
[       OK ] FusedAttentionTest.test_fused_attention_bwd_batch_size=1_seq_len=384_num_heads=1_head_dim=32_causal=False_use_segment_ids=True
[ RUN      ] FusedAttentionTest.test_fused_attention_bwd_batch_size=2_seq_len=384_num_heads=2_head_dim=32_causal=False_use_segment_ids=False
[       OK ] FusedAttentionTest.test_fused_attention_bwd_batch_size=2_seq_len=384_num_heads=2_head_dim=32_causal=False_use_segment_ids=False
[ RUN      ] FusedAttentionTest.test_fused_attention_bwd_batch_size=2_seq_len=384_num_heads=2_head_dim=32_causal=False_use_segment_ids=True
[       OK ] FusedAttentionTest.test_fused_attention_bwd_batch_size=2_seq_len=384_num_heads=2_head_dim=32_causal=False_use_segment_ids=True
[ RUN      ] FusedAttentionTest.test_fused_attention_fwd_batch_size=1_seq_len=384_num_heads=1_head_dim=64_causal=False_use_fwd=False_use_segment_ids=False_kwargs={}
[       OK ] FusedAttentionTest.test_fused_attention_fwd_batch_size=1_seq_len=384_num_heads=1_head_dim=64_causal=False_use_fwd=False_use_segment_ids=False_kwargs={}
[ RUN      ] FusedAttentionTest.test_fused_attention_fwd_batch_size=1_seq_len=384_num_heads=1_head_dim=64_causal=False_use_fwd=False_use_segment_ids=True_kwargs={}
[       OK ] FusedAttentionTest.test_fused_attention_fwd_batch_size=1_seq_len=384_num_heads=1_head_dim=64_causal=False_use_fwd=False_use_segment_ids=True_kwargs={}
[ RUN      ] FusedAttentionTest.test_fused_attention_fwd_batch_size=1_seq_len=384_num_heads=1_head_dim=64_causal=True_use_fwd=False_use_segment_ids=True_kwargs={}
[       OK ] FusedAttentionTest.test_fused_attention_fwd_batch_size=1_seq_len=384_num_heads=1_head_dim=64_causal=True_use_fwd=False_use_segment_ids=True_kwargs={}
[ RUN      ] FusedAttentionTest.test_fused_attention_fwd_batch_size=1_seq_len=384_num_heads=8_head_dim=64_causal=True_use_fwd=False_use_segment_ids=False_kwargs={'block_q': 128, 'block_k': 64}
[       OK ] FusedAttentionTest.test_fused_attention_fwd_batch_size=1_seq_len=384_num_heads=8_head_dim=64_causal=True_use_fwd=False_use_segment_ids=False_kwargs={'block_q': 128, 'block_k': 64}
[ RUN      ] FusedAttentionTest.test_fused_attention_fwd_batch_size=1_seq_len=384_num_heads=8_head_dim=64_causal=True_use_fwd=True_use_segment_ids=False_kwargs={}
[       OK ] FusedAttentionTest.test_fused_attention_fwd_batch_size=1_seq_len=384_num_heads=8_head_dim=64_causal=True_use_fwd=True_use_segment_ids=False_kwargs={}
[ RUN      ] FusedAttentionTest.test_fused_attention_fwd_batch_size=1_seq_len=384_num_heads=8_head_dim=64_causal=True_use_fwd=True_use_segment_ids=True_kwargs={}
[       OK ] FusedAttentionTest.test_fused_attention_fwd_batch_size=1_seq_len=384_num_heads=8_head_dim=64_causal=True_use_fwd=True_use_segment_ids=True_kwargs={}
[ RUN      ] FusedAttentionTest.test_fused_attention_fwd_batch_size=2_seq_len=384_num_heads=2_head_dim=64_causal=False_use_fwd=False_use_segment_ids=True_kwargs={}
[       OK ] FusedAttentionTest.test_fused_attention_fwd_batch_size=2_seq_len=384_num_heads=2_head_dim=64_causal=False_use_fwd=False_use_segment_ids=True_kwargs={}
[ RUN      ] FusedAttentionTest.test_fused_attention_fwd_batch_size=2_seq_len=384_num_heads=8_head_dim=64_causal=True_use_fwd=True_use_segment_ids=True_kwargs={}
[       OK ] FusedAttentionTest.test_fused_attention_fwd_batch_size=2_seq_len=384_num_heads=8_head_dim=64_causal=True_use_fwd=True_use_segment_ids=True_kwargs={}
```

**Comment by sharadmv** [comment]
One last thing, could you try it with seqlen=1024  and a small batch size (1 or 2)?

**Comment by luyug** [comment]
Done.

```
[ RUN      ] FusedAttentionTest.test_fused_attention_bwd_batch_size=1_seq_len=384_num_heads=1_head_dim=32_causal=False_use_segment_ids=False                                               
I0112 22:10:22.682980 140693078705728 xla_bridge.py:670] Unable to initialize backend 'rocm': NOT_FOUND: Could not find registered platform with name: "rocm". Available platform names are:
 CUDA                                                                                                                                                                                      
I0112 22:10:22.684721 140693078705728 xla_bridge.py:670] Unable to initialize backend 'tpu': INTERNAL: Failed to open libtpu.so: libtpu.so: cannot open shared object file: No such file or
directory                                                                                                                                                                                  
[       OK ] FusedAttentionTest.test_fused_attention_bwd_batch_size=1_seq_len=384_num_heads=1_head_dim=32_causal=False_use_segment_ids=False
[ RUN      ] FusedAttentionTest.test_fused_attention_bwd_batch_size=1_seq_len=384_num_heads=1_head_dim=32_causal=False_use_segment_ids=True                                                 
[       OK ] FusedAttentionTest.test_fused_attention_bwd_batch_size=1_seq_len=384_num_heads=1_head_dim=32_causal=False_use_segment_ids=True                                                 
[ RUN      ] FusedAttentionTest.test_fused_attention_bwd_batch_size=2_seq_len=1024_num_heads=2_head_dim=32_causal=False_use_segment_ids=False                                               
[       OK ] FusedAttentionTest.test_fused_attention_bwd_batch_size=2_seq_len=1024_num_heads=2_head_dim=32_causal=False_use_segment_ids=False                                               
[ RUN      ] FusedAttentionTest.test_fused_attention_bwd_batch_size=2_seq_len=384_num_heads=2_head_dim=32_causal=False_use_segment_ids=False                                                
[       OK ] FusedAttentionTest.test_fused_attention_bwd_batch_size=2_seq_len=384_num_heads=2_head_dim=32_causal=False_use_segment_ids=False                                                
[ RUN      ] FusedAttentionTest.test_fused_attention_bwd_batch_size=2_seq_len=384_num_heads=2_head_dim=32_causal=False_use_segment_ids=True                                                 
[       OK ] FusedAttentionTest.test_fused_attention_bwd_batch_size=2_seq_len=384_num_heads=2_head_dim=32_causal=False_use_segment_ids=True 
```

(I did find this long seq test as well as the first test failing without removing the casting line.)

**Review by sharadmv — APPROVED** [review]

**Comment by sharadmv** [comment]
Thank you!

**Comment by superbobry** [comment]
This logic seems to be commented out at head. Closing.
